### PR TITLE
Auth Feature

### LINKED ISSUE #3: feature/Add the Authenticaton

### PR DESCRIPTION
**Pull request by ali-akan**

### CODE REVIEW

File: src/firebase/auth.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  sendEmailVerification,
+  sendPasswordResetEmail,
+  signInWithEmailAndPassword,
+  signInWithPopup,
+  updatePassword,
+} from "firebase/auth";
+
+export const doCreateUserWithEmailAndPassword = async (email, password) => {
+  return createUserWithEmailAndPassword(auth, email, password);
+};
```
Comment (bcakko) [inline src/firebase/auth.js]: Can we change the user account related function names in auth domain: remove "do" in the beginning and add "auth" prefix? 
```suggestion
export const authCreateUser = async (email, password) => {
  return createUserWithEmailAndPassword(auth, email, password);
};
```

File: src/firebase/firebase.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
+  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
+  appId: process.env.REACT_APP_FIREBASE_APP_ID,
+  measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENT_ID,
+};
+
+// Initialize Firebase
+const app = initializeApp(firebaseConfig);
+const auth = getAuth(app);
+const db = getDatabase(app);
```
Comment (bcakko) [inline src/firebase/firebase.js]:
```suggestion
const firebaseApp = initializeApp(firebaseConfig);
const firebaseAuth = getAuth(app);
const firebaseDb = getDatabase(app);
```

File: src/firebase/firebase.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+export { app, auth, db };
```
Comment (bcakko) [inline src/firebase/firebase.js]:
```suggestion
export { firebaseApp, firebaseAuth, firebaseDb };
```

File: src/context/authContext/index.jsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    if (user) {
+      setCurrentUser({ ...user });
+      setUserLoggedIn(true);
+    } else {
+      setCurrentUser(null);
+      setUserLoggedIn(false);
+    }
+    setLoading(false);
+  }
+  const value = {
```
Comment (bcakko) [inline src/context/authContext/index.jsx]: Add a space here

File: src/context/authContext/index.jsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    currentUser,
+    userLoggedIn,
+    loading,
+  };
+
+  return (
+    <AuthContext.Provider value={value}>
+      {!loading && children}
+    </AuthContext.Provider>
+  );
```
Comment (bcakko) [inline src/context/authContext/index.jsx]: Instead of blocking whole app from rendering, can we navigate the user to the relevant page when they're logged in? They can get navigated to auth page if they're not logged in for example. Therefore "loading" state can be removed here.

File: src/context/authContext/index.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import { useContext, useEffect, useState } from "react";
+
+const AuthContext = React.createContext();
+
+export function useAuth() {
+  return useContext(AuthContext);
+}
+
+export function AuthProvider({ children }) {
+  const [currentUser, setCurrentUser] = useState(null);
```
Comment (bcakko) [inline src/context/authContext/index.jsx]: What info does this state contain? Please update the variable names accordingly. If it's a token, it can be named userToken for example.

File: src/context/authContext/index.jsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  const [userLoggedIn, setUserLoggedIn] = useState(false);
```
Comment (bcakko) [inline src/context/authContext/index.jsx]: This state can be renamed as isLoggedIn.

**Review by bcakko — CHANGES_REQUESTED** [review]
Generally looks good. Only needs some detail changes. Can you also add the actual authentication page with login and register form? This addition can contain react-query api integration and use of mutations.

File: src/context/authContext/index.jsx (hunk 5 above)
Comment (ali-akan) [reply src/context/authContext/index.jsx]: To resolve this, useNavigate has been employed.

File: src/context/authContext/index.jsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import { useNavigate } from "react-router-dom";
+
+const AuthContext = React.createContext();
+
+export function useAuth() {
+  return useContext(AuthContext);
+}
+
+export function AuthProvider({ children }) {
+  const [userToken, setUserToken] = useState(null); // Changed from currentUser to userToken
```
Comment (bcakko) [inline src/context/authContext/index.jsx]: Can you change this to userInfo, and remove the comment?
```suggestion
  const [userInfo, setUserInfo] = useState(null);
```

File: src/pages/Home.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { useAuth } from "../context/authContext";
+import { Box } from "@mui/material";
+
+const Home = () => {
+  const { userToken } = useAuth();
```
Comment (bcakko) [inline src/pages/Home.js]: Also userInfo here.

File: src/pages/Home.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  return (
+    <Box>
+      <div>
+        Hello {userToken.displayName ? userToken.displayName : userToken.email},
+        you are now logged in.
+      </div>
+    </Box>
```
Comment (bcakko) [inline src/pages/Home.js]: Use Typography from MUI.

File: src/pages/Login.js
Comment (bcakko) [inline src/pages/Login.js:1]: Can we add react-hook-form here? And maybe create an input component to render in the form. And please replace every HTML element with MUI components.

File: src/pages/Login.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    if (!googleSignInMutation.isLoading) {
+      googleSignInMutation.mutate().catch((err) => {
+        console.error("Error signing in with Google:", err);
+      });
+    }
+  };
+
+  return (
+    <div>
+      {userLoggedIn && <Navigate to={"/home"} replace={true} />}
```
Comment (bcakko) [inline src/pages/Login.js]: You can achieve this with a simple if condition in the component.
`if (!isLoggedIn) { navigate('/home') }`

File: src/pages/Register.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      }
+    }
+  };
+  
+  
+  
+
+  return (
+    <>
+      {userLoggedIn && <Navigate to={"/home"} replace={true} />}
```
Comment (bcakko) [inline src/pages/Register.js]: You can achieve this also with an if condition in the component. Please use navigate function and 
`if (!isLoggedIn) {
    navigate('/home')
}`

**Review by bcakko — CHANGES_REQUESTED** [review]
Looks good, still have some missing points. Can you make use of MUI and react-hook-form, replace every HTML element with MUI components?

File: src/pages/Register.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
+import React from "react";
+import { Navigate, Link } from "react-router-dom";
+import { useAuth } from "../context/authContext/index";
+import { useMutation } from "react-query"; // Import QueryClient and QueryClientProvider
+import { authCreateUser } from "../firebase/auth";
+
+const Register = () => {
+  const { userLoggedIn } = useAuth();
```
Comment (bcakko) [inline src/pages/Register.js]: I believe this should be `isLoggedIn`

File: src/components/Starting Content/Header.js
Comment (bcakko) [inline src/components/Starting Content/Header.js:1]: This is not a component. Please move this under /src/features/Header/index.tsx

File: src/components/Starting Content/Header.js [hunk 14]
```diff
@@ -1,57 +1,37 @@
...
+import { Button } from "@mui/material";
 
 const Header = () => {
-  // const [value, setValue] = React.useState(0);
-
   const navigate = useNavigate();
   const { isLoggedIn } = useAuth();
+
   return (
     <nav>
```
Comment (bcakko) [inline src/components/Starting Content/Header.js]: Please replace it with a Box from MUI?
`<Box component="nav" ... />`

File: src/components/Starting Content/Header.js [hunk 15]
```diff
@@ -1,57 +1,37 @@
...
       {isLoggedIn ? (
         <div>
-          <button
+          <Button
             onClick={() => {
               authSignOut().then(() => {
                 navigate("/login");
               });
             }}
```
Comment (bcakko) [inline src/components/Starting Content/Header.js]: Can you make this a function and use react-query for the async operation?

File: src/components/Starting Content/Header.js [hunk 16]
```diff
@@ -1,57 +1,37 @@
...
 const Header = () => {
-  // const [value, setValue] = React.useState(0);
-
   const navigate = useNavigate();
   const { isLoggedIn } = useAuth();
+
   return (
     <nav>
       {isLoggedIn ? (
         <div>
```
Comment (bcakko) [inline src/components/Starting Content/Header.js]: Please use fragment.

File: src/components/Starting Content/Header.js [hunk 17]
```diff
@@ -1,57 +1,37 @@
...
               });
             }}
+            variant="outlined"
           >
             Logout
-          </button>
+          </Button>
         </div>
       ) : (
         <div>
```
Comment (bcakko) [inline src/components/Starting Content/Header.js]: Also fragment here.

File: src/context/authContext/index.jsx [hunk 18]
```diff
@@ -11,7 +11,7 @@ export function useAuth() {
 }
 
 export function AuthProvider({ children }) {
-  const [userToken, setUserToken] = useState(null); // Changed from currentUser to userToken
+  const [userInfo, setUserInfo] = useState(false);
```
Comment (bcakko) [inline src/context/authContext/index.jsx]: Can this be boolean? I'd set it to `null` for initialization.

File: src/pages/Home.js [hunk 19]
```diff
@@ -1,14 +1,15 @@
 import { useAuth } from "../context/authContext";
 import { Box } from "@mui/material";
+import Typography from "@mui/material/Typography";
 
 const Home = () => {
-  const { userToken } = useAuth();
+  const { userInfo } = useAuth();
   return (
     <Box>
```
Comment (bcakko) [inline src/pages/Home.js]: Also fragment here.

File: src/pages/Login.js [hunk 20]
```diff
@@ -33,58 +44,75 @@ const Login = () => {
...
-      {userLoggedIn && <Navigate to={"/home"} replace={true} />}
-      <div>
-        <form onSubmit={onSubmit}>
-          <div>
-            <label>Email</label>
-            <input type="email" name="email" autoComplete="email" required />
-          </div>
+    <>
+      {isLoggedIn ? (
+        <Navigate to={"/home"} replace={true} />
```
Comment (bcakko) [inline src/pages/Login.js]: Can you move this check under an `if` condition and remove it from jsx?

File: src/pages/Login.js [hunk 21]
```diff
@@ -33,58 +44,75 @@ const Login = () => {
...
+        <div>
           <div>
-            <label>Password</label>
-            <input
-              type="password"
-              name="password"
-              autoComplete="current-password"
-              required
-            />
+            <form onSubmit={handleSubmit(onSubmit)}>
```
Comment (bcakko) [inline src/pages/Login.js]: `<Box component="form" ... />`

File: src/pages/Login.js [hunk 22]
```diff
@@ -33,58 +44,75 @@ const Login = () => {
...
+              <div>
```
Comment (bcakko) [inline src/pages/Login.js]: Fragment

File: src/pages/Login.js [hunk 23]
```diff
@@ -33,58 +44,75 @@ const Login = () => {
...
+                  label="Email"
+                  type="email"
+                  {...register("email", { required: true })}
+                  autoComplete="email"
+                  error={!!errors.email}
+                  helperText={errors.email && "Email is required"}
+                  required
+                />
+              </div>
+              <div>
```
Comment (bcakko) [inline src/pages/Login.js]: Fragment

File: src/pages/Register.js [hunk 24]
```diff
@@ -1,93 +1,133 @@
...
-  const { userLoggedIn } = useAuth();
+  const { isLoggedIn } = useAuth();
+  const {
+    register,
+    handleSubmit,
+    watch,
+    formState: { errors },
+  } = useForm();
+  const password = React.useRef({});
+  password.current = watch("password", "");
```
Comment (bcakko) [inline src/pages/Register.js]: You can listen the value directly with `watch` by providing the registered name to the method. 
`const password = watch("password");`

File: src/pages/Register.js [hunk 25]
```diff
@@ -1,93 +1,133 @@
...
-  
-  
 
   return (
     <>
-      {userLoggedIn && <Navigate to={"/home"} replace={true} />}
-      <main>
-        <div>
+      {isLoggedIn ? (
+        <Navigate to={"/home"} replace={true} />
```
Comment (bcakko) [inline src/pages/Register.js]: Also move this under an `if` condition.

File: src/pages/Register.js [hunk 26]
```diff
@@ -1,93 +1,133 @@
...
-            <div>
-              <label>Confirm Password</label>
-              <input
-                type="password"
-                name="confirmPassword"
-                autoComplete="off"
-                required
-              />
-            </div>
+              <div>
```
Comment (bcakko) [inline src/pages/Register.js]: Fragment

File: src/pages/Register.js [hunk 27]
```diff
@@ -1,93 +1,133 @@
...
+                }}
+              >
+                {registerMutation.isLoading ? (
+                  <CircularProgress size={24} color="inherit" />
+                ) : (
+                  "Sign Up"
+                )}
+              </Button>
+
+              <div>
```
Comment (bcakko) [inline src/pages/Register.js]: Typography

File: src/theme/theme.js [hunk 28]
```diff
@@ -13,7 +13,9 @@ const theme = createTheme({
       primary: "#f9ea0e",
       secondary: "rgba(0,0,0,0.54)",
     },
+  
   },
+
```
Comment (bcakko) [inline src/theme/theme.js]: you can remove these extra lines

**Review by bcakko — CHANGES_REQUESTED** [review]
Looks good, thanks for the changes. There are only some code quality changes needed.

File: src/features/Header/Header.js [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+
+  const signOutMutation = useMutation(authSignOut, {
+    onSuccess: () => {
+      navigate("/login");
+    },
+  });
+
+  const handleSignOut = () => {
+    signOutMutation.mutate();
+  };
```
Comment (bcakko) [inline src/features/Header/Header.js]: Good use of mutations. The only thing could be destructuring the mutate function of useMutation.

```suggestion
  const { mutate: signOut } = useMutation(authSignOut, {
    onSuccess: () => {
      navigate("/login");
    },
  });

  const handleSignOut = () => {
    signOut();
  };
```

File: src/features/Header/Header.js [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+
+  return (
+    <Box>
```
Comment (bcakko) [inline src/features/Header/Header.js]: You don't need a wrapper here.

File: src/features/Header/Header.js [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+      {isLoggedIn ? (
+        <Box>
```
Comment (bcakko) [inline src/features/Header/Header.js]: Fragment

File: src/features/Header/Header.js [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+          <Button
+            onClick={handleSignOut}
+            variant="outlined"
+            disabled={signOutMutation.isLoading}
+          >
+            {signOutMutation.isLoading ? "Logging out..." : "Logout"}
+          </Button>
+        </Box>
+      ) : (
+        <Box>
```
Comment (bcakko) [inline src/features/Header/Header.js]: Also fragment here.

File: src/index.js [hunk 33]
```diff
@@ -5,12 +5,19 @@ import "./index.css";
 import { BrowserRouter } from "react-router-dom";
 import { ThemeProvider } from "@mui/material";
 import theme from "./theme/theme";
+
```
Comment (bcakko) [inline src/index.js:8]: This extra line can be removed.

File: src/pages/Home.js [hunk 34]
```diff
@@ -1,5 +1,24 @@
...
+      <Typography>
+        {userInfo ? (
+          <span>
+            Hello {userInfo.displayName ? userInfo.displayName : userInfo.email}
+            , you are now logged in.
+          </span>
+        ) : (
+          "Loading..."
+        )}
+      </Typography>
```
Comment (bcakko) [inline src/pages/Home.js]: Can you make `userInfo.displayName` and `userInfo.email ` constants, and remove `span`?

File: src/pages/Login.js [hunk 35]
```diff
@@ -1,5 +1,122 @@
...
+    handleSubmit,
+    formState: { errors },
+  } = useForm();
+
+  const signInMutation = useMutation(authSignInWith);
+  const googleSignInMutation = useMutation(authSignInWithGoogle);
+
+  const onSubmit = async (data) => {
+    const { email, password } = data;
+    if (!signInMutation.isLoading) {
```
Comment (bcakko) [inline src/pages/Login.js]: You can also destructure `isLoading` from the mutation. Please only do this if you also make the destructuring I mentioned in previous comments.

File: src/pages/Login.js [hunk 36]
```diff
@@ -1,5 +1,122 @@
...
-  return <div>This is a Login page Test!</div>;
+  const { isLoggedIn } = useAuth();
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+  } = useForm();
+
+  const signInMutation = useMutation(authSignInWith);
+  const googleSignInMutation = useMutation(authSignInWithGoogle);
```
Comment (bcakko) [inline src/pages/Login.js]: You can use `onError` of mutation configuration.

File: src/pages/Login.js [hunk 37]
```diff
@@ -1,5 +1,122 @@
...
+
+  const onSubmit = async (data) => {
+    const { email, password } = data;
+    if (!signInMutation.isLoading) {
+      try {
+        await signInMutation.mutateAsync({ email, password });
+      } catch (error) {
+        console.error("Error signing in:", error);
+      }
```
Comment (bcakko) [inline src/pages/Login.js]: onError

File: src/pages/Login.js (hunk 20 above)
Comment (bcakko) [reply src/pages/Login.js]: This one isn't resolved. Can you move this under an `if` check?

File: src/pages/Login.js [hunk 38]
```diff
@@ -1,5 +1,122 @@
...
+  };
+
+  return (
+    <>
+      {isLoggedIn ? (
+        <Navigate to={"/home"} replace={true} />
+      ) : (
+        <Box>
+          <Box component="form" onSubmit={handleSubmit(onSubmit)}>
+            {" "}
```
Comment (bcakko) [inline src/pages/Login.js]: Please remove this space.

File: src/pages/Login.js [hunk 39]
```diff
@@ -1,5 +1,122 @@
...
+      });
+    }
+  };
+
+  return (
+    <>
+      {isLoggedIn ? (
+        <Navigate to={"/home"} replace={true} />
+      ) : (
+        <Box>
```
Comment (bcakko) [inline src/pages/Login.js]: Is `Box` required here? Is it used for styling?

File: src/pages/Login.js [hunk 40]
```diff
@@ -1,5 +1,122 @@
...
+          <Box component="form" onSubmit={handleSubmit(onSubmit)}>
+            {" "}
+            {/* Use Box as a form .  Since FormControl component does not have an onSubmit prop.*/}
+            <FormControl>
+              <Box>
```
Comment (bcakko) [inline src/pages/Login.js]: Is `Box` required here?

File: src/pages/Register.js [hunk 41]
```diff
@@ -1,5 +1,132 @@
...
+  };
+
+  if (isLoggedIn) {
+    return <Navigate to={"/home"} replace={true} />;
+  }
+
+  return (
+    <Box>
+      <Box>
+        <Box>
```
Comment (bcakko) [inline src/pages/Register.js]: ?

**Review by bcakko — CHANGES_REQUESTED** [review]
Good use of mutations. Can still be improved with a few touches.

File: src/features/Header/Header.js (hunk 29 above)
Comment (bcakko) [reply src/features/Header/Header.js]: You won't be needing handleSignOut function anymore because it basically returns the mutate function. You should be able to call `signOut` directly where you called `handleSignOut`.

File: src/index.js (hunk 33 above)
Comment (ali-akan) [reply src/index.js:8]: index.css stands for body (margin:0, font-family etc,) .

File: src/features/Header/Header.js [hunk 42]
```diff
@@ -0,0 +1,49 @@
...
+  return (
+    <>
+      {isLoggedIn ? (
+        <>
+          <Button
+            onClick={handleSignOut}
+            variant="outlined"
+            disabled={signOut.isLoading}
+          >
+            {signOut.isLoading ? "Logging out..." : "Logout"}
```
Comment (bcakko) [inline src/features/Header/Header.js]: You can destructure also the isLoading parameter from the mutation.

```suggestion
  const { mutate: signOut, isLoading: isSignOutLoading } = useMutation(authSignOut, {
    onSuccess: () => {
      navigate("/login");
    },
  });

  const handleSignOut = () => {
    signOut();
  };

  return (
    <>
      {isLoggedIn ? (
        <>
          <Button
            onClick={handleSignOut}
            variant="outlined"
            disabled={signOut.isLoading}
          >
            {isSignOutLoading ? "Logging out..." : "Logout"}
```

File: src/pages/Home.js [hunk 43]
```diff
@@ -1,5 +1,20 @@
...
+  const email = userInfo?.email;
+
+  return (
+    <Typography>
+      {userInfo ? (
+        <>Hello {displayName || email}, you are now logged in.</>
+      ) : (
+        "Loading..."
+      )}
+    </Typography>
```
Comment (bcakko) [inline src/pages/Home.js]:
```suggestion
  const email = userInfo?.email;
  const loginMessage = userInfo ?
        `Hello ${displayName || email}, you are now logged in.`
       : 
        "Loading...";

  return <Typography>{loginMessage}</Typography>
```

File: src/pages/Login.js [hunk 44]
```diff
@@ -1,5 +1,119 @@
...
+  });
+
+  const onSubmit = async (data) => {
+    const { email, password } = data;
+    try {
+      await signInMutation.mutateAsync({ email, password });
+    } catch (error) {
+      console.error("Error signing in:", error);
+    }
+  };
```
Comment (bcakko) [inline src/pages/Login.js]: Do you need an async operation here? You already handled the error case in the mutation configuration. Try catch doesn't look necessary. Also destructuring the mutate function can look better.

```suggestion
  const onSubmit = (data) => {
    const { email, password } = data;
    signInMutation.mutateAsync({ email, password });
  };
```

File: src/pages/Register.js [hunk 45]
```diff
@@ -1,5 +1,128 @@
...
+
+  if (isLoggedIn) {
+    return <Navigate to={"/home"} replace={true} />;
+  }
+
+  return (
+    <>
+      <Box>
+        <Box>
+          <Typography variant="h6">Create a New Account</Typography>
```
Comment (bcakko) [inline src/pages/Register.js]: Why is this required?

**Review by bcakko — CHANGES_REQUESTED** [review]
Thanks for the changes, I think we can merge after a few small adjustments.

File: src/pages/Login.js (hunk 44 above)
Comment (ali-akan) [reply src/pages/Login.js:44]: const onSubmit = (data) => {
    const { email, password } = data;
    signInMutation.mutateAsync({ email, password });
  };

File: src/pages/Home.js [hunk 46]
```diff
@@ -5,16 +5,11 @@ const Home = () => {
   const { userInfo } = useAuth();
   const displayName = userInfo?.displayName;
   const email = userInfo?.email;
+  const loginMessage = userInfo
+    ? `Hello ${displayName || email}, you are now logged in.`
+    : "Loading...";
```
Comment (bcakko) [inline src/pages/Home.js:10]: Please double check if this is throwing an error in the console.

**Review by bcakko — APPROVED** [review]
Overall looks very good. Thanks for the changes 👍
